doc/protocol.md

```diff
@@ -0,0 +1,259 @@
+the Rationale section for discussion of this design decision.  To resume from a
+failure, a `resume` message, along with a checkpoint object is sent to the
+server. 
+
+The frequency of checkpoint messages can be requested by the client using the
+`X-GA4GH-CHECKPOINT-FREQUENCY` request header.  This is an advisory request;
+the exact frequency of checkpoints is up to the discretion of the server.  The
+value is the approximate number of bytes to stream before sending
+a checkpoint message. 
+
+The client tunes the checkpoint frequency based on network speed and
+reliability.  Since it's expected that that checkpoint will on be used for
+larger transfers over wide area networks, the default value is disabled (0),
+There is no requirement that clients implement checkpoint resume.
+
+The same facility will be used for implementing checkpoints originating at
+from the client once a write API has been defined by GA4GH.
+
+### Message types
+
+The following message types are defined:
+- `dataType typeKey typeName`: Defines a key used to identify the types of
+  data objects that follow in a stream.  This supports mixed GA4GH data types
+  in given stream. A message for a given type is only sent once, before any
+  data objects of that type are sent.  They may be sent at any point in the
+  stream, the only requirement is that a `dataType` message must precede the
+  first message of the type.  If a stream is resumed from a checkpoint,
+  `dataType` objects will be resent before objects of a given type are sent.
+  The`typeKey` parameter is an integer that is included in the `dataObject`
+  messages.  The value is server-instance dependent and its scope is only the
+  current data stream.  If a request is resumed, it should not be assume that
+  the `typeKey` values will be the same as the original stream.  The
+  `typeName` parameter is a full qualified record name `package.message` that
+  identifies the type.  All GA4GH objects will match `org.ga4gh.*`.  Non-GA4GH
+  data types may also use this protocol, however they must not be under the
+  `org.ga4gh` package.
+
+- `dataObject typeKey data` - Used to send a data object.  The `typeKey`
+  parameter is used to find the matching `dataType` message.  The `data`
+  parameter is the data object.
+
+- `checkpoint memento` - Specify a checkpoint in the data stream.
+  The `memento` is an opaque, server-dependent object.  It should be
+  returned as-is when resuming a stream.
+
+- `resume memento` - Resume a transfer at the specified checkpoint, as define
+  by the `memento` object, in data stream.  The next data object that was or
+  would have been received after  the corresponding `checkpoint` message
+  will be the next data object received on the stream. It will be preceded by
+  the required `dataType` message.  The same checkpoint may resumed multiple
+  times, as would be required if there was a failure before another
+  `checkpoint` messages is received in the stream.
```

Comment:
Good point and I don't know the answer.  Maybe retention is
policy?  Obviously, you can't keep them forever.

This goes back to the goals and use cases of GA4GH API being
undefined.  Is transferring larger amounts of data over WANs a
goal?



Jerome Kelleher <notifications@github.com> writes:
> In doc/protocol.md:
> 
> > +
> > +- `dataObject typeKey data` - Used to send a data object.  The `typeKey`
> > +  parameter is used to find the matching `dataType` message.  The `data`
> > +  parameter is the data object.
> > +
> > +- `checkpoint memento` - Specify a checkpoint in the data stream.
> > +  The `memento` is an opaque, server-dependent object.  It should be
> > +  returned as-is when resuming a stream.
> > +
> > +- `resume memento` - Resume a transfer at the specified checkpoint, as define
> > +  by the `memento` object, in data stream.  The next data object that was or
> > +  would have been received after  the corresponding `checkpoint` message
> > +  will be the next data object received on the stream. It will be preceded by
> > +  the required `dataType` message.  The same checkpoint may resumed multiple
> > +  times, as would be required if there was a failure before another
> > +  `checkpoint` messages is received in the stream.
> 
> This is slightly worrying. Does this mean that the server must remember every
> checkpoint that it issues, ever? The server must be able to forget about
> checkpoints at some stage I think.
> 
> —
> Reply to this email directly or view it on GitHub.*
>